shadowsocks/stream_test.go

```diff
@@ -173,3 +176,240 @@ func TestEndToEnd(t *testing.T) {
 		t.Fatalf("Expected output '%v'. Got '%v'", expected, output.String())
 	}
 }
+
+func TestLazyWriteFlush(t *testing.T) {
+	cipher := newTestCipher(t)
+	buf := new(bytes.Buffer)
+	writer := NewShadowsocksWriter(buf, cipher)
+	header := []byte{1, 2, 3, 4}
+	n, err := writer.LazyWrite(header)
+	if n != 4 {
+		t.Errorf("Wrong write size: %d", n)
+	}
+	if err != nil {
+		t.Errorf("LazyWrite failed: %v", err)
+	}
+	if buf.Len() != 0 {
+		t.Errorf("LazyWrite isn't lazy: %v", buf.Bytes())
+	}
+	if err = writer.Flush(); err != nil {
+		t.Errorf("Flush failed: %v", err)
+	}
+	len1 := buf.Len()
+	if len1 == 0 {
+		t.Errorf("No bytes flushed")
+	}
+
+	// Check that normal writes now work
+	body := []byte{5, 6, 7, 8}
+	n, err = writer.Write(body)
+	if n != 4 {
+		t.Errorf("Wrong write size: %d", n)
+	}
+	if err != nil {
+		t.Errorf("Write failed: %v", err)
+	}
+	if buf.Len() == len1 {
```

Comment:
Also add a check that buf.Len() == 8 to show the first write got flushed

Reply:
I also suggest you make the second write be 3 or 5 bytes, so it's not confused with the first write.

Reply:
OK, done